usb: cdc_acm: Fix usb host-reset issue

metratec-mkoehler opened the pull request:
Fixes issue #76064

cdc_acm_irq_tx_ready() and cdc_acm_irq_is_pending() now take configured and suspended states into account
like cdc_acm_fifo_fill() already does.

Tested with the described reproduction from the issue multiple times.

tmon-nordic commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -624,7 +624,8 @@ static int cdc_acm_irq_tx_ready(const struct device *dev)
 
-	if (dev_data->tx_irq_ena && dev_data->tx_ready) {
+	if (dev_data->tx_irq_ena && dev_data->tx_ready && dev_data->configured &&
+	    !dev_data->suspended) {
```
This is touching upon a much bigger issue that never was properly addressed in Zephyr. The problem has been happening multiple times and the workaround was to disable remote wakeup via Kconfig so the device did not claim to support remote wakeup even if the device controller does.

One peculiar feature of USB suspend is that device being suspended doesn't mean that host is suspended. In fact, Linux feature called "autosuspend" is enabled by default. The way autosuspend works with CDC ACM is that if there is no data for host to send and no data is being received for autosuspend duration (e.g. 2 seconds) and **IF** the device supports Remote Wakeup, the device will be suspended. The perfectly fair expectation is that if the device has anything for host to read, then the device will initiate Remote Wakeup. Zephyr CDC ACM implementation doesn't and so we have multiple reports hitting exactly the same issue in various scenarios...

One might think than Remote Wakeup is not proper to do on "just some random log", but this problem is already solved outside Zephyr. The key is that host can disable the Remote Wakeup feature before suspending the device. If the device is suspended with disabled Remote Wakeup, then likely host is actually suspended and the user does not want the CDC ACM to be able to wakeup the system. What devices can or cannot wakeup the system from "full system suspend" is (host) system administration task and can be done e.g. with udev rules on Linux.

Therefore, in order to properly support suspend, the device code has to be aware of Remote Wakeup and issue Remote Wakeup if there is something to be sent. In order for this to happen the chicken-and-egg problem has to be solved where the rest of system should receive TX ready while USB device is suspended.

metratec-mkoehler replied:
Very insightful, thank you

The issue in my case lies in configured, not suspended, so I think at least for the issue at hand just adding the configured check would do. 

Still, currently cdc_acm_fifo_fill() rejects data while suspended so cdc_acm_irq_tx_ready() should reflect that. I don't feel competent to change the way the suspend is handled. A change on this part must be done by someone else in a different pull request.

(I could see cdc_acm_fifo_fill() using 
`if (!cdc_acm_irq_tx_ready())`
as the check instead of 
`if (!dev_data->configured || dev_data->suspended)` so a change to the way suspend works would automatically be updated. Also, cdc_acm_fifo_fill() currently doesn't check if tx_ready is true and it would add that check, too.)

metratec-mkoehler replied:
I also noticed cdc_acm_fifo_fill() can take negative len values which will have unexpected results and isn't checked.

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -624,7 +624,8 @@ static int cdc_acm_irq_tx_ready(const struct device *dev)
 
-	if (dev_data->tx_irq_ena && dev_data->tx_ready) {
+	if (dev_data->tx_irq_ena && dev_data->tx_ready && dev_data->configured &&
+	    !dev_data->suspended) {
```
I already mentioned in another PR review that these changes cannot go in as they are. Also, did you read my comment about [issue](https://github.com/zephyrproject-rtos/zephyr/issues/76064#issuecomment-2237742867) you reported? What fixed it here for you is just the additional check for being configured, but not being configured should not be a reason to not trigger the TX interrupt as long as there is space in the ring buffer.

metratec-mkoehler replied:
I personally prefer data to be rejected when suspended / not configured, but that's not my decision.

My issue is: 
cdc_acm_irq_tx_ready()
and
cdc_acm_fifo_fill()
need to match. When cdc_acm_irq_tx_ready() is true cdc_acm_fifo_fill() needs to take at least one byte. Everything else violates the 
[UART API](https://docs.zephyrproject.org/2.7.5/reference/peripherals/uart.html#group__uart__interface_1ga5e126b5f19549eb7f5b785b98ebe7638).

> Check if UART TX buffer can accept at least one character for transmission (i.e. [uart_fifo_fill()](https://docs.zephyrproject.org/2.7.5/reference/peripherals/uart.html#group__uart__interface_1gafe42e4719eada7e25904bc9ebfe87791) will succeed and return non-zero).


So either cdc_acm_irq_tx_ready() changes as I proposed (matching cdc_acm_fifo_fill() not taking data) OR cdc_acm_fifo_fill() changes taking the data even though it can't be sent (as not configured / suspended). As I said, cdc_acm_fifo_fill() does it right in my opinion, but I am not willing to fight about it. Just let me fix the API issue whichever way you prefer.

As stated (https://github.com/zephyrproject-rtos/zephyr/pull/76171#discussion_r1687512260) I would use cdc_acm_irq_tx_ready() for the check in cdc_acm_fifo_fill() so this way it always matches even if the implementation of suspend gets changed later (or, even cleaner, make another (inline?) function called by both).

metratec-mkoehler commented:
Changed commit message

- removed the fixes info
- changed info from issue centric view it problem centric (UART API issue)
- Added the proposed changes using a single inline function
- Added a buffer space check as this is also expected

metratec-mkoehler committed:
usb: cdc_acm: Fix UART API

cdc_acm_irq_tx_ready() now uses _fifo_fill_ready() to predict
cdc_acm_fifo_fill() result as expected by API.

cdc_acm_fifo_fill() uses _fifo_fill_ready(), too, for consitency
and ease of reading.

cdc_acm_irq_is_pending() now uses cdc_acm_irq_tx_ready() to inherit
the same fix (and for consitency, cdc_acm_irq_is_pending() claims
to check exaclty this).

Signed-off-by: Martin Koehler <koehler@metratec.com>

metratec-mkoehler commented:
Rebased to merge log change

mmahadevan108 commented:
@jfischer-no , can you help take a look at this PR one more time with the updates made to address review comments.

jfischer-no commented:
> @jfischer-no , can you help take a look at this PR one more time with the updates made to address review comments.

Where do you see it being addressed or updated?

metratec-mkoehler commented:
@jfischer-no   

My [commit](https://github.com/zephyrproject-rtos/zephyr/pull/76171/commits/4aecd016eef5c4ed34b964b4ec2a7df9fa878432)  2 weeks ago

I narrowed the issue to it's core (UART API violation) and changed the commit description to also be clear about what it fixes.

metratec-mkoehler commented:
I see my fixes were used for a commit and pushed 2 days ago without noticing me, this pr or naming my work on the topic or the issue they fix. 
I will look into it if all parts of the issue is fixed.
If everything is fixed I'll close the pull request and issue.

I am not sure what to say to this.
It's kinda helpful but feels really rude. 
I'm glad it's likely over.

Have a nice weekend everyone.

tmon-nordic commented:
> I see my fixes were used for a commit and pushed 2 days ago without noticing me, this pr or naming my work on the topic or the issue they fix.

Can you point where this occurred?

metratec-mkoehler commented:
Commit c5a31aec3c6acd2e3d7d4f34dec97d2961a3eb72
cdc_acm_irq_rx_ready() and cdc_acm_irq_is_pending() changes

Commit b79ffe35820cf50c52195f37465723b5e1097921
cdc_acm_irq_rx_ready() issue fixed by cdc_acm_irq_update() change

I don't care about the code (it's mostly trivial stuff) but about not noticing this PR and the issue. And therefor also for ignoring the work on finding the source of the issue. I understand there's frustration because it took so long but I am frustrated about that as well, we need the fix mainlined, too. Noticing this only by luck ... not nice.
